assets/src/edit-story/app/story/storyProvider.js

```diff
@@ -6,98 +6,111 @@ import PropTypes from 'prop-types';
-	const [ currentPageNumber, setCurrentPageNumber ] = useState( null );
-	const [ currentPage, setCurrentPage ] = useState( null );
-	const [ selectedElements, setSelectedElements ] = useState( [] );
+	// Generate selection info
+	const {
+		selectedElementIds,
+		selectedElements,
+		hasSelection,
+	} = useMemo(
+		() => {
+			if ( ! currentPage ) {
+				return {
+					selectedElements: [],
+					selectedElementIds: [],
+					hasSelection: false,
+				};
+			}
+			const els = currentPage.elements.filter( ( { id } ) => selection.includes( id ) );
+			return {
+				selectedElementIds: selection,
+				selectedElements: els,
+				hasSelection: els.length > 0,
+			};
+		},
+		[ currentPage, selection ],
+	);
 
-	const hasSelection = Boolean( selectedElementIds.length );
+	// This effect loads and initialises the story on first load (when there's no pages).
+	const shouldLoad = pages.length === 0;
+	useLoadStory( { restore, shouldLoad, storyId } );
 
-	const clearSelection = useClearSelection( { selectedElementIds, setSelectedElementIds } );
-	const deleteSelectedElements = useDeleteSelectedElements( { currentPageIndex, pages, selectedElementIds, setPages, setSelectedElementIds } );
-	const setCurrentPageByIndex = useSetCurrentPageByIndex( { clearSelection, setCurrentPageIndex } );
-	const addBlankPage = useAddBlankPage( { pages, setPages, clearSelection } );
-	const deleteCurrentPage = useDeleteCurrentPage( { currentPage, pages, setPages, addBlankPage, setCurrentPageIndex, currentPageIndex } );
-	const selectElementById = useSelectElementById( { setSelectedElementIds } );
-	const toggleElementIdInSelection = useToggleElementIdInSelection( { selectedElementIds, setSelectedElementIds } );
-	const appendElementToCurrentPage = useAppendElementToCurrentPage( { currentPageIndex, pages, setPages, setSelectedElementIds } );
-	const setPropertiesOnSelectedElements = useSetPropertiesOnSelectedElements( { currentPageIndex, pages, selectedElementIds, setPages } );
-	const savePost = useSavePost( { isSaving, storyId, title, postStatus, postAuthor, slug, pages, setLink, setPostStatus, setIsSaving } );
-	const setPropertiesById = useSetPropertiesById( { currentPageIndex, pages, setPages } );
+	// These effects send updates to and restores state from history.
+	useHistoryEntry( { pages, current, selection, story } );
+	useHistoryReplay( { restore } );
 
-	useLoadStory( { storyId, pages, setPages, setTitle, setPostStatus, setPostAuthor, setSlug, setLink, setCurrentPageIndex, clearSelection } );
-	useCurrentPage( { currentPageIndex, pages, setCurrentPage, setCurrentPageNumber } );
-	useHistoryEntry( { currentPageIndex, pages, selectedElementIds } );
-	useHistoryReplay( { setCurrentPageIndex, setPages, setSelectedElementIds } );
-	useSelectedElements( { currentPageIndex, pages, selectedElementIds, setSelectedElements } );
+	// This action allows the user to save the story
+	// (and it will have side-effects, thus the need for `updateStory`)
```

Comment:
Q: What do you mean by side-effects here?

Reply:
Saving a post in WP can [change its status and link](https://github.com/ampproject/amp-wp/pull/3910/files#diff-5157e5359482c52319bc0f4b51134219R76-R79) and that'll update the state.

Reply:
Could we specify that in the comment about the side-effects?
Also, is this something that should be mentioned in `useSaveStory` as well?

Reply:
It's pretty obvious inside the function, as the side-effect is right there in the code, but it'll update the description.

Reply:
Fixed in 3d70ba8.